src/controller/user.ts

```diff
@@ -0,0 +1,214 @@
+        userToBeSaved.dob = context.request.body.dob
+        userToBeSaved.following = []
+        userToBeSaved.address = context.request.body.address || ''
+        userToBeSaved.description = context.request.body.description || ''
+
+        const validationErrors = await validate(userToBeSaved, userSchema, ['name', 'email', 'dob', 'password'])
+
+        if (validationErrors) return response(context, 400, validationErrors)
+
+        const call = await safeCall(userRepository.findOne({ email: userToBeSaved.email }))
+
+        if (call.err) return response(context, 500, call.err)
+
+        if (call.res) return response(context, 409, 'user_already_exists')
+
+        // convert users dob to date object
+        if (userToBeSaved.dob) userToBeSaved.dob = new Date(userToBeSaved.dob)
+
+        try {
+            await userToBeSaved.hashPassword()
+            await userRepository.save(userToBeSaved)
+        } catch {
+            return response(context, 400, 'error_creating_user')
+        }
+
+        response(context, 201, 'user_created')
+    }
+
+    @request('patch', '/users/{id}')
+    @summary('Update a user')
+    @path({ id: { type: 'string', required: true, description: 'id of the user to update' }, data: { type: 'object', items: {} } })
+    @body(_.omit(userSchema.properties, ['id', 'createdAt', 'updatedAt', 'following']))
+    @responses({
+        200: { description: 'user updated successfully' },
+        400: { description: 'user not found, user already exists, validation errors, user already exists' },
+        401: { description: 'token authorization error' },
+        403: { description: 'not authorized to perform this action' },
+    })
+    public static async updateUser(context: BaseContext): Promise<void> {
+        const userRepository: Repository<User> = getManager().getRepository(User)
+
+        const userToBeUpdated: User = new User()
+
+        userToBeUpdated.id = context.params.id
+        userToBeUpdated.name = context.request.body.name
+        userToBeUpdated.email = context.request.body.email
+        userToBeUpdated.dob = context.request.body.dob
+        userToBeUpdated.address = context.request.body.address
+        userToBeUpdated.description = context.request.body.description
+        delete userToBeUpdated.createdAt
+
+        const validationErrors = await validate(userToBeUpdated, userSchema, ['id'])
+
+        if (validationErrors) return response(context, 400, validationErrors)
+
+        let call = await safeCall(userRepository.findOne(userToBeUpdated.id))
+
+        if (call.err) return response(context, 500, call.err)
+
+        if (!call.res) return response(context, 400, 'user_not_found')
+
+        call = await safeCall(
+            userRepository.findOne({ id: Not(Equal(userToBeUpdated.id)), email: userToBeUpdated.email }),
+        )
+
+        if (call.err) return response(context, 500, call.err)
+
+        if (call.res) return response(context, 400, 'user_already_exists')
```

Comment:
```if (call.err) return response(context, 500, call.err)```

Any chance this can leak something we don't want to expose?

```
if (call.res) return response(context, 400, 'user_already_exists')
```

Here the response body is a string. Mixing content-type in API does not look like a good practice to me.

Reply:
At worst this will return an stringed Error message indicating database connection error since all format validation is handled before `safeCall` makes a database request and this function only wraps database calls. But you are right, it is better to not mix content-type and return `server_error` or `database_connection_error` depending on the needs of the receiving application.